### PR TITLE
kube-apiserver add BuiltInAuthorizationOptions validation

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Validate BuiltInAuthorizationOptions after flags parsed. 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@deads2k ptal

File: pkg/kubeapiserver/authorizer/config.go [hunk 1]
```diff
@@ -129,14 +113,6 @@ func (config AuthorizationConfig) New() (authorizer.Authorizer, authorizer.RuleR
...
-	}
-
-	if !authorizerMap[modes.ModeABAC] && config.PolicyFile != "" {
-		return nil, nil, errors.New("Cannot specify --authorization-policy-file without mode ABAC")
-	}
-	if !authorizerMap[modes.ModeWebhook] && config.WebhookConfigFile != "" {
-		return nil, nil, errors.New("Cannot specify --authorization-webhook-config-file without mode Webhook")
 	}
 
 	return union.New(authorizers...), union.NewRuleResolvers(ruleResolvers...), nil
```
Comment (deads2k) [inline pkg/kubeapiserver/authorizer/config.go]: If you have no authorizer at this point, you must fail.  If you return without failing, you run the risk of running the cluster without any authorization.

File: pkg/kubeapiserver/options/authorization.go [hunk 2]
```diff
@@ -38,19 +40,57 @@ type BuiltInAuthorizationOptions struct {
 
 func NewBuiltInAuthorizationOptions() *BuiltInAuthorizationOptions {
 	return &BuiltInAuthorizationOptions{
-		Mode: authzmodes.ModeAlwaysAllow,
+		Modes: []string{authzmodes.ModeAlwaysAllow},
```
Comment (deads2k) [inline pkg/kubeapiserver/options/authorization.go]: Do we actually want this to be the default?  I don't think so.  Is it even still the default for kube-apiserver?  I thought we switched to RBAC.

File: pkg/kubeapiserver/options/authorization.go [hunk 3]
```diff
@@ -38,19 +40,57 @@ type BuiltInAuthorizationOptions struct {
...
 }
 
 func (s *BuiltInAuthorizationOptions) Validate() []error {
+	if s == nil {
+		return nil
+	}
 	allErrors := []error{}
+
+	if len(s.Modes) == 0 {
+		allErrors = append(allErrors, fmt.Errorf("at least one authorization-mode should be passed"))
```
Comment (deads2k) [inline pkg/kubeapiserver/options/authorization.go]: must be passed

File: pkg/kubeapiserver/options/authorization.go [hunk 4]
```diff
@@ -38,19 +40,57 @@ type BuiltInAuthorizationOptions struct {
...
+	if len(s.Modes) != len(modes.List()) {
+		allErrors = append(allErrors, fmt.Errorf("authorization-mode %q has mode specified more than once", s.Modes))
+	}
+
 	return allErrors
 }
 
 func (s *BuiltInAuthorizationOptions) AddFlags(fs *pflag.FlagSet) {
-	fs.StringVar(&s.Mode, "authorization-mode", s.Mode, ""+
+	fs.StringSliceVar(&s.Modes, "authorization-mode", s.Modes, ""+
```
Comment (deads2k) [inline pkg/kubeapiserver/options/authorization.go]: Need a test to make sure this still works with commas.  As I recall it, we changed cobra a while back to stop splitting on commas by default.

**Comment by deads2k** [comment]
I like it in concept.

File: pkg/kubeapiserver/options/authorization.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/kubeapiserver/options/authorization.go]: I think this maybe reserved for test

File: pkg/kubeapiserver/authorizer/config.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/kubeapiserver/authorizer/config.go]: make sense.  But have already checked in `BuiltInAuthorizationOptions.Validate`

File: pkg/kubeapiserver/options/authorization.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/kubeapiserver/options/authorization.go]: ok

File: pkg/kubeapiserver/options/authorization.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pkg/kubeapiserver/options/authorization.go]: Yes, I will add test in `cmd/kube-apiserver/app/options/options_test.go`
In theory,   as `enable-admission-plugins` fs.StringSliceVar(&a.EnablePlugins, "enable-admission-plugins",  it must work well with commas

File: pkg/kubeapiserver/authorizer/config.go (hunk 1 above)
Comment (deads2k) [reply pkg/kubeapiserver/authorizer/config.go]: > make sense. But have already checked in BuiltInAuthorizationOptions.Validate

Relying on functions to be called in order is poor design.  For preventing panics or ugly errors, calling functions in order is more acceptable.  For preventing behavioral failures, it is not.  You need to test for empty.

Comment (hzxuzhonghu) [reply pkg/kubeapiserver/authorizer/config.go]: ic, learned a lot

**Comment by hzxuzhonghu** [comment]
@deads2k comments addressed

File: pkg/kubeapiserver/authorizer/config.go (hunk 1 above)
Comment (deads2k) [reply pkg/kubeapiserver/authorizer/config.go]: This comment was resolved at beginning of function

File: pkg/kubeapiserver/options/authorization.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/kubeapiserver/options/authorization.go]: ok

**Comment by hzxuzhonghu** [comment]
squashed

**Pull request merged**